src/scheduler/data_types.h

```diff
@@ -773,6 +781,27 @@ struct group_info
 	group_info *child;			/* child node */
 };
 
+/**
+ * Set of equivalent resresvs.  It is used to keep track if one can't run, the rest cannot.
+ * The set is defined by a number of attributes of the resresv.  If the attributes do
+ * not matter, they won't be used and set to NULL
+ */
+struct resresv_set
```

Comment:
The logic isn't with sinfo->has_user_limit, it's actually in a function that looks at that element.  I put in an @see create_resresv_set_by_resresv() as a place to look for why the members can be NULL.
I dislike the state idea.  I chose can_not_run due to its analogous bit on the resource_resv structure.  Can you see other uses of state?